File: android/app/src/main/java/com/bwqr/mavinote/models/Note.kt
package com.bwqr.mavinote.models

import com.bwqr.mavinote.reax.deserializeOption
import com.novi.serde.DeserializationError
import com.novi.serde.Deserializer

data class Note constructor(
    val id: Int,
    val folderId: Int,
    val remoteId: Int?,
    val commit: Int,
    val name: String,
    val text: String,
    val state: State
) {
    companion object {
        fun deserialize(deserializer: Deserializer): Note {
            deserializer.increase_container_depth()

            val note = Note(
                deserializer.deserialize_i32(),
                deserializer.deserialize_i32(),
                deserializeOption(deserializer) { it.deserialize_i32() },
                deserializer.deserialize_i32(),
                deserializer.deserialize_str(),
                deserializer.deserialize_str(),
                State.deserialize(deserializer)
            )

            deserializer.decrease_container_depth()

            return note
        }
    }
}

enum class State {
    Clean,
    Modified,
    Deleted;

    companion object {
        fun deserialize(deserializer: Deserializer): State {
            val index = deserializer.deserialize_variant_index()

            return when (index) {
                0 -> Clean
                1 -> Modified
                2 -> Deleted
                else -> throw DeserializationError("Unknown variant index for State: $index")
            }
        }
    }
}

File: ios/Mavinote/Models/Note.swift
import Serde

struct Note : Identifiable, Deserialize {
    let id: Int32
    let folderId: Int32
    let remoteId: Int32?
    let commit: Int32
    let name: String
    let text: String
    let state: ModelState

    static func deserialize(_ deserializer: Deserializer) throws -> Note {
        try deserializer.increase_container_depth()

        let note = Note(
            id: try deserializer.deserialize_i32(),
            folderId: try deserializer.deserialize_i32(),
            remoteId: try De.Option<De.I32>.deserialize(deserializer),
            commit: try deserializer.deserialize_i32(),
            name: try deserializer.deserialize_str(),
            text: try deserializer.deserialize_str(),
            state: try ModelState.deserialize(deserializer)
        )

        try deserializer.decrease_container_depth()

        return note
    }
}

enum ModelState: Deserialize {
    case Clean
    case Modified
    case Deleted

    static func deserialize(_ deserializer: Deserializer) throws -> ModelState {
        try deserializer.increase_container_depth()

        let index = try deserializer.deserialize_variant_index()

        let state: ModelState

        switch index {
        case 0: state = Clean
        case 1: state = Modified
        case 2: state = Deleted
        default: throw DeserializationError.invalidInput(issue: "Invalid variant for State \(index)")
        }

        try deserializer.decrease_container_depth()

        return state
    }

}
